src/hotspot/cpu/aarch64/aarch64_vector_ad.m4

```diff
@@ -1870,56 +1870,119 @@ REDUCE_ADD_FP_PREDICATE(D, D)
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this, $vsrc);
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this, $vsrc);
+    __ reduce_mul_integral_gt128b($dst$$Register, bt, $isrc$$Register,
+                                  $vsrc$$FloatRegister, length_in_bytes,
+                                  $tmp1$$FloatRegister, $tmp2$$FloatRegister,
+                                  $pgtmp$$PRegister);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct reduce_mulL_le128b(iRegLNoSp dst, iRegL isrc, vReg vsrc) %{
   predicate(Matcher::vector_length_in_bytes(n->in(2)) == 16);
   match(Set dst (MulReductionVL isrc vsrc));
   effect(TEMP_DEF dst);
-  format %{ "reduce_mulL $dst, $isrc, $vsrc\t# 2L" %}
+  format %{ "reduce_mulL_le128b $dst, $isrc, $vsrc\t# 2L" %}
+  ins_encode %{
+    __ reduce_mul_integral_le128b($dst$$Register, T_LONG, $isrc$$Register,
+                                  $vsrc$$FloatRegister, 16, fnoreg, fnoreg);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct reduce_mulL_gt128b(iRegLNoSp dst, iRegL isrc, vReg vsrc, vReg tmp1,
+                            pRegGov pgtmp) %{
+  predicate(Matcher::vector_length_in_bytes(n->in(2)) > 16);
+  match(Set dst (MulReductionVL isrc vsrc));
+  effect(TEMP_DEF dst, TEMP tmp1, TEMP pgtmp);
+  format %{ "reduce_mulL_gt128b $dst, $isrc, $vsrc\t# vector (> 128 bits). KILL $tmp1, $pgtmp" %}
   ins_encode %{
-    __ neon_reduce_mul_integral($dst$$Register, T_LONG, $isrc$$Register,
-                                $vsrc$$FloatRegister, 16, fnoreg, fnoreg);
+    uint length_in_bytes = Matcher::vector_length_in_bytes(this, $vsrc);
+    __ reduce_mul_integral_gt128b($dst$$Register, T_LONG, $isrc$$Register,
+                                  $vsrc$$FloatRegister, length_in_bytes,
+                                  $tmp1$$FloatRegister, fnoreg,
+                                  $pgtmp$$PRegister);
   %}
   ins_pipe(pipe_slow);
 %}
 
-instruct reduce_mulF(vRegF dst, vRegF fsrc, vReg vsrc, vReg tmp) %{
+instruct reduce_mulF_le128b(vRegF dst, vRegF fsrc, vReg vsrc, vReg tmp) %{
   predicate(Matcher::vector_length_in_bytes(n->in(2)) <= 16);
   match(Set dst (MulReductionVF fsrc vsrc));
   effect(TEMP_DEF dst, TEMP tmp);
-  format %{ "reduce_mulF $dst, $fsrc, $vsrc\t# 2F/4F. KILL $tmp" %}
+  format %{ "reduce_mulF_le128b $dst, $fsrc, $vsrc\t# 2F/4F. KILL $tmp" %}
   ins_encode %{
     uint length_in_bytes = Matcher::vector_length_in_bytes(this, $vsrc);
-    __ neon_reduce_mul_fp($dst$$FloatRegister, T_FLOAT, $fsrc$$FloatRegister,
-                          $vsrc$$FloatRegister, length_in_bytes, $tmp$$FloatRegister);
+    __ reduce_mul_fp_le128b($dst$$FloatRegister, T_FLOAT, $fsrc$$FloatRegister,
+                            $vsrc$$FloatRegister, length_in_bytes, $tmp$$FloatRegister);
   %}
   ins_pipe(pipe_slow);
 %}
 
-instruct reduce_mulD(vRegD dst, vRegD dsrc, vReg vsrc, vReg tmp) %{
```

Comment:
Please consider that `reduce_mulF_gt128b` and `reduce_mulD_gt128b` might be similar enough that they should be combined in the same way as other patterns in this file.